pygem/cad/ffd.py

```diff
@@ -0,0 +1,353 @@
+        direction. If not provided it is set to [2, 2, 2].
+
+    :cvar numpy.ndarray box_length: dimension of the FFD bounding box, in the
+        x, y and z direction (local coordinate system).
+    :cvar numpy.ndarray box_origin: the x, y and z coordinates of the origin of
+        the FFD bounding box.
+    :cvar numpy.ndarray rot_angle: rotation angle around x, y and z axis of the
+        FFD bounding box.
+    :cvar numpy.ndarray n_control_points: the number of control points in the
+        x, y, and z direction.
+    :cvar numpy.ndarray array_mu_x: collects the displacements (weights) along
+        x, normalized with the box length x.
+    :cvar numpy.ndarray array_mu_y: collects the displacements (weights) along
+        y, normalized with the box length y.
+    :cvar numpy.ndarray array_mu_z: collects the displacements (weights) along
+        z, normalized with the box length z.
+
+    :Example:
+
+        >>> import pygem.freeform as ffd
+        >>> import pygem.params as ffdp
+        >>> import numpy as np
+        >>> ffd = FFD()
+        >>> ffd.read_parameters(
+        >>>        'tests/test_datasets/parameters_test_ffd_iges.prm')
+        >>> input_cad_file_name = "input.iges"
+        >>> modified_cad_file_name = "output.iges"
+        >>> ffd(input_cad_file_name,modified_cad_file_name)
+    """
+    def __init__(self, n_control_points=None):
+        super().__init__(n_control_points=None)
+        self.uKnotsToAdd = 30
+        self.vKnotsToAdd = 30
+        self.knotsToAdd = 30
+        self.tolerance = 1e-4
+
+    def __call__(self, obj, dst=None):
+        """
+        This method performs the deformation on the CAD file.
+        """
+
+        # Manage input
+        if isinstance(obj, str): # if a input filename is passed
+            iges_handler = IgesHandler()
+            shape = iges_handler.load_shape_from_file(obj)
+        elif isinstance(obj, TopoDS_Shape):
+            shape = obj
+        # Maybe do we need to handle also Compound?
+        else:
+            raise TypeError
+
+
+        #create compound to store modified faces
+        compound_builder = BRep_Builder()
+        compound = TopoDS_Compound()
+        compound_builder.MakeCompound(compound)
+
+
+        # cycle on the faces to get the control points
+        # init some quantities
+        faceCount = 0
+        face_list = []
+        faces_explorer = TopExp_Explorer(shape, TopAbs_FACE)
+
+        while faces_explorer.More():
+            # performing some conversions to get the right
+            # format (BSplineSurface)
+            face = topods_Face(faces_explorer.Current())
+            nurbs_converter = BRepBuilderAPI_NurbsConvert(face)
+            nurbs_face = nurbs_converter.Shape()
+            face_aux = topods_Face(nurbs_face)
+            brep_face = BRep_Tool.Surface(topods_Face(nurbs_face))
+            bounds = 0.0
+            bounds = brep_face.Bounds()
```

Comment:
Class private method (one underscore in front of the name e.g. `_foo()`) for performing the cast. Describe input and output variables.